### PR TITLE
Add multirow section

### LINKED ISSUE #2169: Multirow section
[Multirow figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A19272&t=1xUFkrsxqomXObMe-1)

- [ ] Settings design for section and blocks / content update on theme settings [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A19272&t=1xUFkrsxqomXObMe-1)
- [ ] Empty state aspect [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=431%3A38870&t=1xUFkrsxqomXObMe-1)

Settings details :
- [ ] Image height [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A23545&t=1xUFkrsxqomXObMe-1)
- [ ] Desktop image width [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A24114&t=1xUFkrsxqomXObMe-1)
- [ ] Heading, text and button styles [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A27394&t=1xUFkrsxqomXObMe-1)
- [ ] Desktop content position  [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A23809&t=1xUFkrsxqomXObMe-1)
- [ ] Desktop content alignment / Mobile content alignement [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A23640&t=1xUFkrsxqomXObMe-1)
- [ ] Desktop image layout (Not mandatory for launch) [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=431%3A35781&t=1xUFkrsxqomXObMe-1)
- [ ] Color scheme [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A24659&t=1xUFkrsxqomXObMe-1)
- [ ] Container color scheme [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A25128&t=1xUFkrsxqomXObMe-1)
- [ ] Usecase if both color options have the same gradient scheme [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=379%3A28848&t=1xUFkrsxqomXObMe-1)
- [ ] Section paddings [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A24403&t=1xUFkrsxqomXObMe-1)
- [ ] Spacing between rows [Figma](https://www.figma.com/file/hQiYDulTbQCKxBqIbtCpWZ/Multirow-and-Shopify-hosted-video-explorations?node-id=372%3A27803&t=1xUFkrsxqomXObMe-1)

@kmeleta

### PR DESCRIPTION
**Pull request by kmeleta**
### PR Summary: 

Added a new Multirow section for displaying related image and text content vertically.

### Why are these changes introduced?

Multirow is a section conceptually similar to `multicolumn`, but designed specifically for quicker and easier vertical content enumeration. Its blocks are aesthetically similar to the current `image with text` section.

Fixes #2169 

### What approach did you take?

Copied the liquid from image with text into a new section, without creating a shared snippet, so much of the markup will be the same. I'm not certain the added snippet complexity that would be required to support both use cases would be worth it. This may be dependent on how each section functionally evolves in the future.

Multirow loops through blocks and creates a new image with text element for each, so the way existing content elements (heading, caption, etc) are read differs a bit from image with text. Certain image with text classes that were conditional on section settings were either removed or hardcoded if the settings are not offered in multirow.

### Other considerations

**Color scheme + global border settings logic** ([Video of behavior](https://screenshot.click/13-49-ajiff-u0u1z.mp4))
I've included the same "magic" as image with text where if the container color scheme chosen matches the section background and _no_ borders or shadow styles are applied via global content container settings. This allows either a [unified container look](https://screenshot.click/13-49-qnvpx-flt77.png) or a [container-less/transparent look](https://screenshot.click/13-48-ncbh5-h0lu0.png). The only difference here is that multirow also allows the option to change the section background so this logic isn't solely linked to background-1

**Shadow stacking limitations** ([Video of behavior](https://screenshot.click/13-30-dsr8l-av93i.mp4))
When vertical shadow offsets for either the global media or content container settings are negative (upward), the shadow could overlap the row above (natural CSS behavior) if there isn't enough gap. The issue here is there are 2 different child elements within an image with text creating their own shadows with an existing stacking relationship of their own, which further adds to the complexity that we've been able to solve for in other sections. I briefly timeboxed a fix for this and didn't find a configuration that worked, but I don't think it's blocking.

**Color scheme + padding logic** ([Image of behavior](https://screenshot.click/03-18-fewou-8xi43.png))
We also aligned to include additional logic to better align the text content with the standard page margins when both color scheme settings are the same and the rows don't create the "container" look. Basically it just removes the appropriate left or right padding depending on the row's content alignment. An exception is also made for centered text.

### Decision log

| # | Decision | Alternatives | Rationale | Downsides |
|---|---|---|---|---|
| 1 | Each block is a row, rather than a single content element | Basically none | - To allow for multiple rows<br>- To allow for connection to dynamic source groups | - Content elements will not be reorder-able like image with text |
| 2 | Block content style settings at the section level | Include all/more content style settings at the block level | - Focus is around quick and easy configurable<br>- More granularity at the block level would be more tedious and time consuming | - All blocks will display with the same setting choices |
| 3 | Did not include _all_ currently available settings from image with text | Include more or all image with text settings | - Focus on settings most relevant to the purpose of this section<br>- Less options, less configuration time for merchants<br>- Easier to add settings later than take some away | - Rows will not have all the same styling options as image with text (e.g overlap layout) |
| 4 | Did not reuse existing translated strings from image with text | Reuse common strings and only add multirow strings when needed | - Allows image with text and multirow to more easily diverge further in the future<br>- Keeps things simpler | - Inherent redundancy |
| 5 | Did not abstract image with text liquid into a reusable snippet | Consolidate core image with text in a settings-agnostic snippet | - Allows image with text and multirow to more easily diverge further in the future<br>- Snippet would need to be completely agnostic of `section.settings` and `block.settings` references resulting in a lot of props<br>- New logic specific to one component will require adding an exception in the shared snippet<br>- I'm not confident if shared liquid here will be beneficial or a burden | - Inherent redundancy<br>- New shared logic will need to be duplicated  |

### Visual impact on existing themes
Multirow is a new section and any changes to image with text styles should be scoped to the multirow context, with one minor exception.. Long words that might not have broken in image with text captions or buttons before should break now ([example](https://camo.githubusercontent.com/72c9f691fc93175f5fc095db289d1f17c17e80672b82fb1ece00ec306309b85a/68747470733a2f2f73637265656e73686f742e636c69636b2f31302d35312d656c646f6d2d6a357767352e706e67))

### Testing steps/scenarios
- Test Multirow sections with various combinations of section settings
- Ensure all settings/options are labeled correctly and behave as expected
- Test color scheme combinations thoroughly
- Test with different combination of global media and content container settings
- Test with dynamic content sources (Puppy product)

~~Note: `Image height: medium` setting may not apply correctly until the PR that adds the styles for this is merged and this one is rebased.~~

### Demo links

- [Editor](https://os2-demo.myshopify.com/admin/themes/139164581910/editor)

### Checklist
- [ ] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

File: assets/component-image-with-text.css
Comment (kmeleta) [inline assets/component-image-with-text.css:330]: > looks like Medium height doesn’t work, as I obtain a super short height.

There's a note about this in the PR description. Medium won't work until https://github.com/Shopify/dawn/pull/2148 merges

> Can we align content on the grid? If that complicated no need to do it.

If we want to introduce the additional magic for it, then technically yes. I just wasn't sure if we were aligned to do that or not. I can add it, but we should probably align as a group on that.

@YoannJailin

Comment (kmeleta) [reply assets/component-image-with-text.css:347]: @YoannJailin padding magic has been added. Intended logic is as follows

Removes..
- left padding on image-second rows
- right padding on image-first rows
- both left/right padding on mobile

If..
- both color scheme settings match, and..
- content container has no visible border, and..
- content container has no visible shadow

File: assets/component-image-with-text.css
Comment (kmeleta) [reply assets/component-image-with-text.css:356]: Far as I know this still isn't possible in native CSS yet (let me know if you've seen differently) and we don't use preprocessors in Dawn

Comment (dan-menard) [reply assets/component-image-with-text.css:356]: oh right 😅 I was thinking of SASS.

File: sections/multirow.liquid
Comment (dan-menard) [reply sections/multirow.liquid:38]: That's fair. Your PR, so ultimately your call 😄 

If there's anything else you think we can do to make the code more reasonable, even if it's like breaking up the code over multiple lines or something, consider it. Otherwise, I'm fine shipping the way it is.

File: sections/multirow.liquid
Comment (dan-menard) [reply sections/multirow.liquid:45]: +1 for other opinions, I'm curious how others feel.

Good idea to consider the rest of the codebase, but we could also use this as an opportunity to start pushing the codebase in a better direction, if we have broad agreement that more abstraction is better.

File: assets/component-image-with-text.css
Comment (kmeleta) [reply assets/component-image-with-text.css:347]: Yeah basically. This is what I'm seeing with the padding removed but text centered. It just looks a little odd to me

File: sections/multirow.liquid
Comment (ludoboludo) [reply sections/multirow.liquid:45]: I wondering if we need the `cycle` ? Could we leverage CSS selectors instead ? `:nth-child(even)` or `:nth-child(odd)`

Comment (ludoboludo) [reply sections/multirow.liquid:45]: 😅 fair. I didn't look too deeply into yet. I was just going through the PR description and comments 😬 Thanks for the context.

File: sections/multirow.liquid
Comment (ludoboludo) [inline sections/multirow.liquid:210]: I think we might need to add a similar setting for the image 🤔 Right now it stretches based on the length of the content. And it seem weird that it'd do that since we have a specific image height setting for it but it doesn't end up respecting it, in this scenario: 
<details>
<summary>Screenshot</summary>


</details>

File: sections/multirow.liquid
Comment (ludoboludo) [inline sections/multirow.liquid:70]: This is going to be a duplicate ID, we could use the `forloop.index` on top of it. `ImageWithText--{{ section.id }}-{{ forloop.index }}`

File: sections/multirow.liquid
Comment (ludoboludo) [inline sections/multirow.liquid:396]: Just for context, [there is a PR](https://github.com/Shopify/dawn/pull/2085) by @kjellr where he changes all headings to be of type `inline_richtext`. He also adds some styling to make sure links are coming up properly.
But depending on who merges first, this might need some attention 👍

File: assets/component-image-with-text.css
Comment (YoannJailin) [reply assets/component-image-with-text.css:347]: @kmeleta I agree! Let's remove the magic on desktop when content is centered then :)

File: sections/multirow.liquid
Comment (kmeleta) [reply sections/multirow.liquid:396]: Ah good note. I'll see about changing this if there's no issues with the dynamic content stuff.

File: assets/component-image-with-text.css
Comment (kmeleta) [reply assets/component-image-with-text.css:353]: > I think here we should maybe not apply that to the last child.

Whoops, did I not handle that? I honestly forget why I decided margins would be simpler. But I can reconsider 👍 

> `<div class="multirow page-width">`

heh, I _really_ don't like this pattern 😄 Even if we weren't adding a section background color now, the possibility that we ever would add that or a fullwidth layout option in the future is enough to include a more robust dom structure from the start. The minimum-viable-markup approach we've historically taken has resulted in a bunch of `multirow-wrapper` type retrofits and generally inconsistent markup between sections.

File: locales/en.default.schema.json
Comment (kmeleta) [reply locales/en.default.schema.json:1440]: These are just what's in the figma currently until we get final copy approval, though my _personal_ preferences are `Alternate from left` and `Align left`.  

I like `Align left` as long as the setting label is called `Image layout`. If the setting was called `Image position` instead I think `Left` would be fine.  cc @YoannJailin

File: sections/multirow.liquid
Comment (kmeleta) [reply sections/multirow.liquid:379]: Right, good question. @YoannJailin do we have an opinion on a block limit?

File: sections/multirow.liquid
Comment (kmeleta) [reply sections/multirow.liquid:305]: Yeah I think I feel that way too. Thoughts on this @YoannJailin ?

File: sections/multirow.liquid
Comment (ludoboludo) [reply sections/multirow.liquid:210]: Yeah that could be something we do a follow up on if we see an actual need for it and if merchants do use longer texts.

File: sections/multirow.liquid
Comment (kmeleta) [reply sections/multirow.liquid:45]: @dan-menard I briefly looked at what abstracting more classes out of line 45 might look like. Abstracting all the global grid-related classes (a common example) ended up looking something like one of these options. Being able to keep the actual dom element's class attr cleaner could really be helpful at times, but I'd probably want to further consider how an approach like this would apply to other sections and types of classes before implementing. That said, any initial reactions from anyone? 
```
{%- capture grid_classes -%}
  grid grid--gapless grid--1-col grid--{% if section.settings.desktop_image_width == 'medium' %}2-col-tablet{% else %}3-col-tablet{% endif %}
{%- endcapture -%}
<div class="image-with-text__grid {{ grid_classes }}{% if section.settings.image_layout contains 'alternate' %}{% cycle odd_class, even_class %}{% else %}{{ odd_class }}{% endif %}">
```
```
{%- liquid
  assign grid_classes = 'grid grid--gapless grid--1-col'
  if section.settings.desktop_image_width == 'medium'
    assign grid_classes = grid_classes | append: 'grid--2-col-tablet'
  else
    assign grid_classes = grid_classes | append: 'grid--3-col-tablet'
  endif
-%}
<div class="image-with-text__grid {{ grid_classes }} {% if section.settings.image_layout contains 'alternate' %}{% cycle odd_class, even_class %}{% else %}{{ odd_class }}{% endif %}">
```

File: assets/component-image-with-text.css
Comment (kmeleta) [reply assets/component-image-with-text.css]: I remember why i didn't use flex. Just that it would require I name the page-width div to style it sensibly. I somewhat want to propose a standard section structure for the theme where this particular element would inherently have a name like `section__inner` (as in `.multirow.section__outer > .section__inner`), but for now I just extended the multirow namespace.

**Review by Rusty-UX — APPROVED** [review]
From the side of content experiences UX this is great! No blocking feedback.

Works really well, and optional captions also work great. Feel free to try it out on [my store](https://rusty-testing-stuff.myshopify.com/admin/themes/122411843697/editor?previewPath=%2Fproducts%2Fvolkl-rtm-8-0-skis-10-0-fastrak-iii-tp-bindings-2016&section=template--14558770954353__ad4a9bcf-eb88-4b6e-9694-693ba637a169)

image

**Review by dan-menard — APPROVED** [review]
👍 🎩  Nice work!

Code looks good to me, and thanks for putting the changes right on the os2 store, that made playing with it a bunch easier 😄

File: sections/multirow.liquid
Comment (melissaperreault) [inline sections/multirow.liquid:226]:
```suggestion
      "default": "middle",
```
This would suit better different length of content and balance better the white space between the rows. It becomes even more obvious with this type of predefined and opinionated layout this section offers.

File: sections/multirow.liquid
Comment (melissaperreault) [inline sections/multirow.liquid:393]: I want to investigate if the `Caption` is intentional to keep by default?! Stay tuned!

File: sections/multirow.liquid
Comment (melissaperreault) [inline sections/multirow.liquid:411]:
```suggestion
          "id": "button_label",
          "default": Button label,
          "label": "t:sections.multirow.blocks.row.settings.button_label.label"
```
I want to investigate if this is the real intention to have the `Button label` default placeholder!? Stay tuned!

File: sections/multirow.liquid
Comment (kmeleta) [reply sections/multirow.liquid:306]: @melissaperreault do you agree with swapping these?

File: sections/multirow.liquid
Comment (melissaperreault) [reply sections/multirow.liquid:380]: No block limit here please, thanks!

**Comment by kmeleta** [comment]
Updated settings with Katy-approved copy (https://github.com/Shopify/dawn/pull/2168/commits/c249130bfa40556f2448651447e14d1dbf1e81bc) @melissaperreault @ludoboludo if one or both of you could verify those I'll request translations.

File: sections/multirow.liquid
Comment (melissaperreault) [reply sections/multirow.liquid:393]: @Rusty-UX If we have the `Caption` label by default visible on the onboarding state of the section, do you have concerns from your perspective?

File: sections/multirow.liquid
Comment (melissaperreault) [inline sections/multirow.liquid:133]:
```suggestion
      "default": "medium",
```
Since the width is set to Medium, I would recommend changing to Medium here to to get closer to a 1:1 aspect ratio.

File: sections/multirow.liquid
Comment (melissaperreault) [reply sections/multirow.liquid:411]: Update: Let's add this for the default button to show up! Thanks!

File: sections/multirow.liquid
Comment (melissaperreault) [inline sections/multirow.liquid:206]: We are good with displaying the button as an outline for this section! 🙌

File: sections/multirow.liquid
Comment (kmeleta) [reply sections/multirow.liquid]: Right, done

**Review by dan-menard — DISMISSED** [review]

File: sections/multirow.liquid
Comment (Rusty-UX) [reply sections/multirow.liquid:393]: @kmeleta @melissaperreault Sorry for the delay - 

👍 **I don't see any issues with that.**

_Context - if interested:_
What will it mean for dynamic sources?
- Will just mean that a merchant will want to clear out that value from `caption` to `empty` if they don't want to use it, which will mirror the case for merchants inputting static values into that setting
- Makes sense to have it so that merchants can see what the caption will look like before they add their own custom data

Comment (melissaperreault) [reply sections/multirow.liquid:393]: Thank you! We are good to go @kmeleta ! 🚀

File: sections/multirow.liquid
Comment (YoannJailin) [reply sections/multirow.liquid:211]: I agree, let's not change the behavior we reused from Image with text. We can correct it on both sections later on.

File: sections/multirow.liquid
Comment (kmeleta) [reply sections/multirow.liquid:77]: I solved this by making the existing image with text word break styles less specific https://github.com/Shopify/dawn/pull/2168/commits/8470fda55b79f8fe0403ad45bb1da5713d4e0907. Not sure I want to attempt to apply these more globally here, though off the top of my head I'm not sure I can think of a scenario we _wouldn't_ want a caption or title to break. What do you think about that?

File: sections/multirow.liquid
Comment (kmeleta) [reply sections/multirow.liquid:81]: mmm good catch. I've just included it in multirow for now, but yeah I wonder.. If we're going to load it in theme.liquid, is it worth just adding the styles to base.css? Probably don't want to consider that type of change here, but maybe I'll capture this in an issue to be fully considered in terms of css loading performance.

File: sections/multirow.liquid
Comment (Rusty-UX) [reply sections/multirow.liquid:399]: Sorry for the delay. 

@kmeleta I don't think we should change anything on naming based on our mapping algorithm. I think it's fine. If heading is the best UX for merchants in editor we should use that. 

_Background context on how mapping works if interested:_
- We check the types first - and only compatible types can be mapped - e.g. file->image field →  image setting (In the case of text we have a few potential compatible settings)
- Then we check the name of the field and the name of the setting, and check their similarity in name, and score that - based on the closeness of the names we match them.

File: sections/multirow.liquid
Comment (melissaperreault) [reply sections/multirow.liquid:77]: Caption and button should break, yes!

File: assets/component-image-with-text.css
Comment (melissaperreault) [reply assets/component-image-with-text.css]: Yeah, I think so. This is improving for when both the container and the website background have the same color and the alignment of text requires a shift to maintain an invisible vertical alignment line. But in a follow up PR, agreed!

File: sections/multirow.liquid
Comment (melissaperreault) [reply sections/multirow.liquid:435]: The intention is to have 3 rows to help merchants build storytelling sections as we observed that most of those layout patterns were made of 3 rows often. We can always revisit as we learn!

File: sections/multirow.liquid
Comment (melissaperreault) [reply sections/multirow.liquid:253]: To align with other sections, the layout specific settings often comes at this level like it is. I think we have plenty to learn from those settings usage regardless. Thanks for this valid feedback and let's keep observing and sharing those thoughts! 🙏

**Review by ludoboludo — APPROVED** [review]
Looking good 👍

**Pull request merged**
